lntest/itest/lnd_test.go

```diff
@@ -4779,6 +4832,57 @@ func testMultiHopPayments(net *lntest.NetworkHarness, t *harnessTest) {
 	closeChannelAndAssert(ctxt, t, net, carol, chanPointCarol, false)
 }
 
+// assertHtlcEvents consumes events from a client and ensures that they are of
+// the expected type and contain the expected number of forwards and settles.
+// This function can be used when payments are sent asynchronously, and can't
+// be processed sequentially.
+func assertHtlcEvents(t *harnessTest, fwdCount, settleCount int,
+	expectedType routerrpc.HtlcEvent_EventType,
+	client routerrpc.Router_SubscribeHtlcEventsClient) {
+
+	var forwards, settles int
+
+	numEvents := fwdCount + settleCount
+	for i := 0; i < numEvents; i++ {
+		event := assertEventAndType(t, expectedType, client)
+
+		switch event.Event.(type) {
```

Comment:
At this point, it is already know that the event is of the given fixed type `expectedType`. Is it still necessary to do this switch? I would think that it is sufficient to have just a single counter, that counts the number of events of that given type.

Reply:
We have two different types we're checking here 
- `assertEventAndType`: check that the htlc event is labelled as a send/receive/forward as expected
- This switch, where we check whether the event is a settle or forward

For example a sending node has a forward of type send, and a settle of type send because those are the two htlc events associated with their single send.